contrib/golang/filters/http/source/go/pkg/http/filtermanager.go

```diff
@@ -63,3 +81,271 @@ func getHttpFilterConfigParser(name string) api.StreamFilterConfigParser {
+			}
+
+			conf.current = append(conf.current, &filterConfig{
+				Name:         proto.Name,
+				parsedConfig: config,
+			})
+		} else {
+			api.LogErrorf("plugin %s not found, ignored", name)
+		}
+	}
+
+	return conf, nil
+}
+
+func (p *filterManagerConfigParser) Merge(parent interface{}, child interface{}) interface{} {
+	// TODO: We have considered to implemented a Merge Policy between the LDS's filter & RDS's per route
+	// config. A thought is to reuse the current Merge method. For example, considering we have
+	// LDS:
+	//	 - name: A
+	//	   pet: cat
+	// RDS:
+	//	 - name: A
+	//	   pet: dog
+	// we will call plugin A's Merge method, which will produce `pet: [cat, dog]` or `pet: dog`.
+	// As there is no real world use case for the Merge feature, I decide to delay its implementation
+	// to avoid premature design.
+	return child
+}
+
+type filterManager struct {
+	filters []api.StreamFilter
+	names   []string
+
+	callbacks api.FilterCallbackHandler
+}
+
+func filterManagerConfigFactory(c interface{}) api.StreamFilterFactory {
+	conf, ok := c.(*filterManagerConfig)
+	if !ok {
+		panic("unexpected config type")
+	}
+
+	newConfig := conf.current
+	factories := make([]api.StreamManagedFilterFactory, len(newConfig))
+	names := make([]string, len(factories))
+	for i, fc := range newConfig {
+		var factory api.StreamManagedFilterConfigFactory
+		name := fc.Name
+		names[i] = name
+		if v, ok := httpManagedFilterConfigFactoryAndParser.Load(name); ok {
+			plugin := v.(*managedFilterConfigFactoryAndParser)
+			factory = plugin.configFactory
+			config := fc.parsedConfig
+			factories[i] = factory(config)
+
+		} else {
+			api.LogErrorf("plugin %s not found, pass through by default", name)
+			factory = ManagedPassThroughFactory
+			factories[i] = factory(nil)
+		}
+	}
+
+	return func(callbacks api.FilterCallbackHandler) api.StreamFilter {
+		filters := make([]api.StreamFilter, len(factories))
+		for i, factory := range factories {
+			filters[i] = factory(callbacks)
+		}
+		return &filterManager{
+			callbacks: callbacks,
+			names:     names,
+			filters:   filters,
+		}
+	}
+}
+
+func (m *filterManager) DecodeHeaders(header api.RequestHeaderMap, endStream bool) api.StatusType {
+	go func() {
+		defer m.callbacks.RecoverPanic()
+
+		for _, f := range m.filters {
+			// When the filter is controlled by filterManager, it don't
+			// need to create goroutine by itself. So there is no need
+			// to return api.Running and resume inside the filter. We
+			// treat api.Running as api.Continue here.
+			status := f.DecodeHeaders(header, endStream)
+			if status == api.LocalReply {
+				return
+			}
+		}
+		m.callbacks.Continue(api.Continue)
+	}()
+
+	return api.Running
+}
+
+func (m *filterManager) notifyUnsupportedStatus(name string, status api.StatusType) (willReturn bool) {
+	if status == api.Running || status == api.Continue {
+		return false
+	}
+
+	// status == StopXXX
+	if len(m.filters) == 1 {
```

Comment:
this may mess up the users?
sometime it works when only one filter used, but sometime it doesn't work when two filters used.